File: Python/0034-find-first-and-last-position-of-element-in-sorted-array.py
# Tags: Array, Binary Search, Medium
# Time: O(nlogn)
# Space: O(1)

class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left, right = 0, len(nums) - 1
        ans1 = ans2 = -1
        while left <= right:
            mid = left + (right - left)//2
            if nums[mid] == target:
                ans1 = mid
                right = mid - 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1

        left, right = 0, len(nums) - 1
        while left <= right:
            mid = left + (right - left)//2
            if nums[mid] == target:
                ans2 = mid
                left = mid + 1
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return [ans1, ans2]

"""
Finding two points of inflection.
"""

File: C++/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
// Tags: Array Binary-search Amazon
// Time: O(nlogn)
// Space: O(1)
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1, ans1 = -1, ans2 = -1;
        // find leftmost
        while (left <= right) {
            int mid = left + (right - left)/2;
            if (nums[mid] == target) {
                ans1 = mid;
                right = mid - 1;
            } else if (nums[mid] > target)
                right = mid - 1;
            else
                left = mid + 1;
        }
        
        left = 0, right = nums.size() - 1;
        // find rightmost
        while (left <= right) {
            int mid = left + (right - left)/2;
            if (nums[mid] == target) {
                ans2 = mid;
                left = mid + 1;
            } else if (nums[mid] > target)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return {ans1, ans2};
    }
};
/* find two points of inflection */
